src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -360,60 +398,52 @@ public JsonPropertyInfo GetProperty(ReadOnlySpan<byte> propertyName, ref ReadSta
 
         public JsonPropertyInfo PolicyProperty { get; private set; }
 
-        private static bool TryIsPropertyRefEqual(in PropertyRef propertyRef, ReadOnlySpan<byte> propertyName, ulong key, ref JsonPropertyInfo info)
-        {
-            if (key == propertyRef.Key)
-            {
-                if (propertyName.Length <= PropertyNameKeyLength ||
-                    // We compare the whole name, although we could skip the first 6 bytes (but it's likely not any faster)
-                    propertyName.SequenceEqual(propertyRef.Info.Name))
-                {
-                    info = propertyRef.Info;
-                    return true;
-                }
-            }
-
-            return false;
-        }
-
-        private static bool IsPropertyRefEqual(ref PropertyRef propertyRef, PropertyRef other)
-        {
-            if (propertyRef.Key == other.Key)
-            {
-                if (propertyRef.Info.Name.Length <= PropertyNameKeyLength ||
-                    propertyRef.Info.Name.AsSpan().SequenceEqual(other.Info.Name.AsSpan()))
-                {
-                    return true;
-                }
-            }
-
-            return false;
-        }
-
+        /// <summary>
+        /// Get a key from the property name.
+        /// If property name length is 7 or less then this key can be used for equality since
+        /// the first 7 bytes are encoded along with the length.
+        /// </summary>
         public static ulong GetKey(ReadOnlySpan<byte> propertyName)
         {
             ulong key;
             int length = propertyName.Length;
 
-            // Embed the propertyName in the first 6 bytes of the key.
-            if (length > 3)
+            // Embed the propertyName in the first 7 bytes of the key.
+            if (length > 7)
+            {
+                key = MemoryMarshal.Read<ulong>(propertyName);
+
+                // Clear the last byte so it can hold the length.
+                // Be consistent by using shift operators instead of & to avoid endianness issues.
+                key = key << 0x8 >> 0x8;
+            }
+            else if (length > 3)
             {
                 key = MemoryMarshal.Read<uint>(propertyName);
-                if (length > 4)
+
+                if (length == 7)
                 {
-                    key |= (ulong)propertyName[4] << 32;
+                    key |= (ulong)propertyName[4] << 0x20
+                        | (ulong)propertyName[5] << 0x28
+                        | (ulong)propertyName[6] << 0x30;
                 }
-                if (length > 5)
+                else if (length == 6)
                 {
-                    key |= (ulong)propertyName[5] << 40;
+                    key |= (ulong)propertyName[4] << 0x20
+                        | (ulong)propertyName[5] << 0x28;
+                }
+                else if (length == 5)
```

Comment:
For code like this, let's make sure to have tests for all cases (i.e. properties of length 1-10).

**Edit:** Spoke too soon. I see the tests :)